proposals/p2188.md

```diff
@@ -0,0 +1,1689 @@
+    case dyn d1: Derived1* => { Print("Derived1"); }
+    case dyn d2: Derived2* => { Print("Derived2"); }
+    default => { Print("Unknown derived class"); }
+  }
+}
+
+fn PrintTemplateType[template T:! Type](p: T*) {
+  match (p) {
+    // OK, dispatch is based on the static type.
+    case b: Base* => { Print("Base"); }
+    case d1: Derived1* => { Print("Derived1"); }
+    case d2: Derived2* => { Print("Derived2"); }
+    default => { Print("Unknown class"); }
+  }
+}
+```
+
+However, at this time we do not have a design for a checked down-cast, so it's
+not clear how this matching operation would fit into the design of classes,
+either syntactically or semantically.
+
+### User-defined pattern matching
+
+We plan to provide a mechanism for allowing a user-defined type to specify how
+it can be matched by patterns. See [proposal #157](/proposals/p0157.md) for
+details.
+
+### Matching classes with struct patterns
+
+We could allow a class to be matched by a struct pattern that matches its
+fields. This would make sense especially for data classes, and would be
+consistent with the behavior of `==` in the case where a struct is implicitly
+convertible to the class type. However, without a design for user-defined
+pattern matching and matching on dynamic type, there is a significant risk that
+this would conflict with the rules there, so it is deferred for now.
+
+### Matching by reference
+
+When the scrutinee is an lvalue, it is sometimes desirable to form a mutable
+binding to it. For example, see the
+[`struct` inspection example](#figure-5-struct-inspection) below. We currently
+support this only for the `me` binding in a method, using the `addr` keyword,
+but could allow this more generally.
+
+```carbon
+fn takeDamage(p: player*) {
+  match (*p) {
+    case {.hitpoints = 0, .lives = 0, _} => { gameOver(); }
+    case {.hitpoints = addr hp: i32*, .lives = addr l: i32*, _} if *hp == 0 => {
+      *hp = 10;
+      --*l;
+    }
+    case {.hitpoints = addr hp: i32*, _} if *hp <= 3 => {
+      --*hp;
+      messageAlmostDead();
+    }
+    case {.hitpoints = addr hp: i32*, _} => {
+      --*hp;
+    }
+  }
+}
+```
+
+Work in this area will need to consider whether we can provide this feature
+ergonomically without introducing reference-like behavior.
+
+### Type deduction
+
+This proposal does not cover type deduction, instead considering it to be a
+separate topic from pattern matching syntax, even though the semantic behavior
+of the two may be quite similar or identical.
+
+We will need a proposal to explore type deduction and describe its functioning.
+
+## Examples
+
+These examples are translations of examples in WG21 paper
+[P0095R1](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0095r1.html)
+and
+[P1371R3](https://www.open-std.org/JTC1/SC22/WG21/docs/papers/2020/p1371r3.pdf),
+with permission from the author of those examples, David Sankel. Thank you,
+David!
+
+### Examples from P0095R1
```

Comment:
Does it make sense to provide a comparison to the syntax proposed in https://wg21.link/p2392 and more generally acknowledge some of the differences in design there?

Reply:
If that comparison is useful when evaluating this proposal, I can add it, but I'm a little hesitant to get too deep into comparing this with in-flight WG21 proposals. I'm a little tempted to remove the P0095R1 / P1371R3 comparisons, given that they're not comparing against an existing language.